### PR TITLE
mrc-1235 add translate directive

### PR DESCRIPTION
**Pull request by hillalex**
As well as translating strings in spans, we sometimes need to translate input attributes, so I added a custom directive that allows for translating arbitrary attributes. 

Now that I've done this I realised we could actually use the directive approach everywhere and it would be terser than the component approach.

`<h4><translated text-key="someKey"></translated></h4>`

becomes

`<h4 v-translate>someKey</h4>`

This PR applies the new directive in the `ForgotPassword.vue` component by way of illustration.
What do you reckon?

### CODE REVIEW

**Comment by EmmaLRussell** [comment]
> As well as translating strings in spans, we sometimes need to translate input attributes, so I added a custom directive that allows for translating arbitrary attributes.
> 
> Now that I've done this I realised we could actually use the directive approach everywhere and it would be terser than the component approach.
> 
> `<h4><translated text-key="someKey"></translated></h4>`
> 
> becomes
> 
> `<h4 v-translate>someKey</h4>`
> 
> This PR applies the new directive in the `ForgotPassword.vue` component by way of illustration.
> What do you reckon?

It would be nice to keep this consistent everywhere rather than maintaining two approaches, and like you say this is a bit terser - feels like it keeps out of the way better..

**Comment by hillalex** [comment]
Yeah I agree. So I'll back out the component and take this directive approach forward.

**Comment by EmmaLRussell** [comment]
Sounds good - do you want to do that work on this branch before merging it?

**Comment by hillalex** [comment]
> Sounds good - do you want to do that work on this branch before merging it?

No I've got some further PRs that build off this one

**Comment by hillalex** [comment]
> Sounds good - do you want to do that work on this branch before merging it?

It's also much easier to test, shouldn't have to amend existing tests at all in fact. 

Go ahead and review this PR first, I've got further PRs that build off this one. It's going to be a lot of changes so trying to batch them.

**Comment by hillalex** [comment]
> Sounds good - do you want to do that work on this branch before merging it?

It's also much easier to test, shouldn't have to amend existing tests at all in fact. 

Go ahead and review this PR first, I've got further PRs that build off this one. It's going to be a lot of changes so trying to batch them.

File: src/app/static/src/app/components/password/ForgotPassword.vue [hunk 1]
```diff
@@ -1,26 +1,34 @@
 <template>
     <div class="card reset-password-form mx-auto mt-5">
         <div class="card-body">
-            <h3 class="card-title">Forgotten your password?</h3>
-            <p>
-                If you've forgotten your password, enter your email address to request a link which you can use
-                to create a new password.
+            <h3 class="card-title" v-translate>
+                forgottenPassword
```
[the remark this answers is not in the document]
Comment (hillalex) [reply src/app/static/src/app/components/password/ForgotPassword.vue]: So I've actually had to introduce exactly this syntax:
`v-translate=keyName` elsewhere now, because `<div v-translate>keyName</div>` only works where `keyName` is static, i.e. `<div v-translate>{{keyName}}</div>` won't work. If you think the former is unintuitive anyway lets just go with the latter everywhere then :+1:

File: src/app/static/src/app/directives/translate.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import i18next from "i18next";
+import {Store} from "vuex";
+import {TranslatableState} from "../root";
+import {DirectiveOptions} from "vue";
+import {Language} from "../store/translations/locales";
+
+export default <S extends TranslatableState>(store: Store<S>): DirectiveOptions => ({
+    bind(el, binding, vnode) {
+
+        (el as any).__key = el.innerText.trim();
```
[the remark this answers is not in the document]
Comment (hillalex) [reply src/app/static/src/app/directives/translate.ts]: `e.__key` is only used for the inner text, not for attributes, so this *should* be ok, but I'll add a test to make sure!

**Comment by hillalex** [comment]
Ok, added tests and extra logic around supporting multiple directives on one element and have re-organised and commented the code as it was getting to a point of complexity that was a little hard to follow.

File: src/app/static/src/tests/directives/translate.test.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        return store as any;
+    };
+
+    it("initialises the attribute with the translated text", () => {
+        const store = createStore();
+        const rendered = shallowMount(TranslateAttributeTest, {store});
+        expect((rendered.find("input").element as HTMLInputElement).value).toBe("Validate");
+    });
+
+    it("translates the attribute when the store language changes", () => {
```
Comment (hillalex) [inline src/app/static/src/tests/directives/translate.test.ts]: Baffled by code cov saying [these lines](https://codecov.io/gh/mrc-ide/hint/compare/602e18c6732cfd57aad8f8fa4dcc496fb9ea2f95...8732beb4ec14949eb2bd5a7c6765bd60e2d67d3c/diff#D8-33) weren't hit, as they're clearly under test, e.g. here

Comment (EmmaLRussell) [reply src/app/static/src/tests/directives/translate.test.ts]: Me too!
